workspaces/marketplace/plugins/marketplace/src/components/CodeEditor.tsx

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+
+import { Progress } from '@backstage/core-components';
+
+import { useTheme } from '@mui/material/styles';
+import Editor, { loader, OnChange, OnMount } from '@monaco-editor/react';
+import * as monaco from 'monaco-editor';
+import type MonacoEditor from 'monaco-editor';
+
+loader.config({ monaco });
+
+interface CodeEditorContextValue {
+  __editor: MonacoEditor.editor.IStandaloneCodeEditor | null;
+  setValue: (value: string) => void;
+  getValue: () => string;
+}
+
+export const CodeEditorContext = React.createContext<CodeEditorContextValue>(
+  undefined as any as CodeEditorContextValue,
+);
+
+export const CodeEditorContextProvider = (props: {
+  children: React.ReactNode;
+}) => {
+  const contextValue = React.useMemo<CodeEditorContextValue>(
+    () => ({
+      __editor: null,
+      setValue: (value: string) => contextValue.__editor?.setValue(value),
+      getValue: () => contextValue.__editor?.getValue() as string,
+    }),
+    [],
+  );
+
+  return (
+    <CodeEditorContext.Provider value={contextValue}>
+      {props.children}
+    </CodeEditorContext.Provider>
+  );
+};
+
+export const useCodeEditor = () => React.useContext(CodeEditorContext);
+
+export interface CodeEditorProps {
+  defaultLanguage: 'yaml'; // We enforce yalm for now
+  defaultValue?: string;
+  onChange?: OnChange;
+  onLoaded?: () => void;
+}
+
+export const CodeEditor = ({
+  defaultLanguage,
+  defaultValue,
+  onChange,
+  onLoaded,
+  ...otherProps
+}: CodeEditorProps) => {
+  const theme = useTheme().palette.mode === 'dark' ? 'vs-dark' : 'vs-light';
+
+  const codeEditor = useCodeEditor();
+
+  const onMount = React.useCallback<OnMount>(
+    (editor, _monaco) => {
+      codeEditor.__editor = editor;
+      onLoaded?.();
+    },
+    [codeEditor, onLoaded],
+  );
+
+  return (
+    <Editor
```

Comment:
Although it's great that we use the monaco editor for syntax highlighting, the YAML support is currently not as complete as it could be. 

We can use the [monaco-yaml](https://www.npmjs.com/package/monaco-yaml/v/3.1.0) package to provide more a comprehensive suite of features, such as JSON-schema based tooltips, validation, etc

Reply:
Thanks, that's really nice option. Currently, the schema is really simple, but yes. We could hopefully improve this later.